provider/sidecar.go

```diff
@@ -123,42 +124,44 @@ func (provider *Sidecar) loadSidecarConfig() error {
 }
 
 func (provider *Sidecar) sidecarWatcher() error {
-	client := &http.Client{Timeout: 0}
+	//set timeout to be just a bot more than connection refresh interval
+	client := &http.Client{Timeout: (provider.RefreshConn + 3) * time.Second}
 	resp, err := client.Get(provider.Endpoint + "/watch")
 	if err != nil {
 		return err
 	}
-	err = catalog.DecodeStream(resp.Body, provider.callbackLoader)
-	if err != nil {
-		return err
-	}
-	for { //check for open http connection, try to re-init if it fails
-		if resp.Close {
-			resp, err = client.Get(provider.Endpoint + "/watch")
-			if err != nil {
-				return err
-			}
-			err = catalog.DecodeStream(resp.Body, provider.callbackLoader)
-			if err != nil {
-				return err
-			}
+	go catalog.DecodeStream(resp.Body, provider.callbackLoader)
+	log.Infof("Using %s Sidecar connection refresh interval", provider.RefreshConn)
+	provider.recycleConn(resp, client)
+	return nil
+}
+
+func (provider *Sidecar) recycleConn(resp *http.Response, client *http.Client) {
+	var err error
+	for { //use refresh interval to occasionally reconnect to Sidecar in case the stream connection is lost
+		time.Sleep(provider.RefreshConn * time.Second)
+		resp, err = client.Get(provider.Endpoint + "/watch")
```

Comment:
That test seems valid. It looks like when an error occurs on the socket, `dec.More()` returns false and then `catalog.DecodeStream()` returns `nil` and the goroutine exits. My mistake! 👍